### PR TITLE
global: sequential responses for the same request INSPIR-1232

### PR DESCRIPTION
**Pull request by ammirate**
This commit allows to build services that reply different responses
to the same request. This is achieved by introducing a `max_replay`
value which tells to the service how many times that interaction
can actually be played again. After this number is reached, the next
matching interaction is used to reply to the same request.

Signed-off-by: Antonio Cesarano <cesarano2607@gmail.com>

### CODE REVIEW

File: tests/integration/test_base_service.py [hunk 1]
```diff
@@ -116,6 +116,36 @@ def test_base_service_process_request_test_scenario_replays_ok(dispatcher):
     assert response_service_2.body == b'test_scenario_replays_ok/TestServiceB/0'
 
 
+def test_base_service_process_request_test_scenario_picks_next_after_reaches_max_replay(dispatcher):
```
Comment (szymonlopaciuk) [inline tests/integration/test_base_service.py]: I would also add a test to check other numbers than `max_replay: 1` for good measure :slightly_smiling_face:

File: inspire_mitmproxy/services/base_service.py [hunk 2]
```diff
@@ -55,9 +55,14 @@ def handles_request(self, request: MITMRequest) -> bool:
 
         return host in self.hosts_list
 
-    def _get_first_matching_interaction(self, request):
+    def _can_replay(self, interaction: Interaction) -> bool:
```
Comment (szymonlopaciuk) [inline inspire_mitmproxy/services/base_service.py]: Technicaly, if the `max_replay: 0` or `1` then it becomes "should play" rather than "replay"

File: inspire_mitmproxy/services/base_service.py [hunk 3]
```diff
@@ -105,6 +110,12 @@ def increment_interaction_count(self, interaction_name: str):
                 {'num_calls': 1},
             )
 
+    def get_interaction_count(self, interaction_name: str) -> int:
```
Comment (szymonlopaciuk) [inline inspire_mitmproxy/services/base_service.py]: (I am so sorry for this :slightly_smiling_face:) Maybe `get_interaction_replay_count` instead? Otherwise could be confused for the number of interactions themselves...

File: inspire_mitmproxy/services/base_service.py (hunk 2 above)
Comment (szymonlopaciuk) [reply inspire_mitmproxy/services/base_service.py]: Also regarding my comment in the integration test, you could instead add a few unit tests to check the corner cases here?

File: inspire_mitmproxy/services/base_service.py [hunk 4]
```diff
@@ -66,7 +71,7 @@ def _raise_do_not_intercept_if_recording(self, request):
 
     def process_request(self, request):
         try:
-            matched_interaction = self._get_first_matching_interaction(request)
+            matched_interaction = self._get_matching_interaction(request)
         except ScenarioNotInService:
```
Comment (szymonlopaciuk) [inline inspire_mitmproxy/services/base_service.py]: Maybe add an indication in the error message that the scenario wasn't necessarily just not found, but maybe it was found but max_replay were exceeded?
